### PR TITLE
Implement Collapser

### PR DESCRIPTION
**Pull request by andrewtarzia**

### CODE REVIEW

File: stko/optimizers/collapser.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+            Defaults to ``True``
+
+        """
+
+        self._output_dir = output_dir
+        self._step_size = step_size
+        self._distance_cut = distance_cut
+        self._scale_steps = scale_steps
+
+    def _get_inter_BB_distance(self, mol):
```
Comment (stevenbennett96) [inline stko/optimizers/collapser.py:93]: I think this should have arg types. I have no idea whether this works for stk molecule or RDKit?

File: stko/optimizers/collapser.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+                    position_matrix=position_matrix,
+                    atom1_id=atom1.get_atom().get_id(),
+                    atom2_id=atom2.get_atom().get_id()
+                )
+                yield dist
+
+    def _get_min_inter_BB_distance(self, mol):
+        """
+        Calculate the minimum inter building block distance.
+
```
[the remark this answers is not in the document]
Comment (stevenbennett96) [reply stko/optimizers/collapser.py:126]: Same as the mol comment above

File: stko/optimizers/collapser.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+            bb_id = atom.get_building_block_id()
+            id = atom.get_atom().get_id()
+            pos = mol.get_position_matrix()[id]
+            new_position_matrix[id] = (
+                pos - step*vectors[bb_id]*scales[bb_id]
+            )
+
+        return new_position_matrix
+
+    def _get_BB_vectors(self, mol, BB_atom_ids):
```
Comment (stevenbennett96) [inline stko/optimizers/collapser.py:164]: I don't understand why some methods have docs but others don't?

File: stko/optimizers/collapser.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+        bb2 = stk.BuildingBlock(
+            smiles='O=CC(C=O)C=O',
+            functional_groups=[stk.AldehydeFactory()],
+        )
+        cage1 = stk.ConstructedMolecule(
+            topology_graph=stk.cage.FourPlusSix((bb1, bb2)),
+        )
+
+        # Perform collapser optimisation.
+        output_dir = f'cage_opt_{cage_name}_coll'
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply stko/optimizers/collapser.py:47]: fixed

File: stko/optimizers/collapser.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+
+        """
+
+        self._output_dir = output_dir
+        self._step_size = step_size
+        self._distance_cut = distance_cut
+        self._scale_steps = scale_steps
+
+    def _get_atom_distance(self, position_matrix, atom1_id, atom2_id):
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply stko/optimizers/collapser.py]: Was done

File: stko/optimizers/collapser.py [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+            v=position_matrix[atom2_id]
+        )
+
+        return float(distance)
+
+    def _get_inter_BB_distance(self, mol):
+        """
+        Yield inter building block distances.
+
+        Ignores H atoms.
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply stko/optimizers/collapser.py:98]: Not all functions or methods require thorough doc strings when their usage is obvious

File: stko/optimizers/collapser.py (hunk 1 above)
Comment (andrewtarzia) [reply stko/optimizers/collapser.py:94]: Handle in the optimize function and class docs.

File: stko/optimizers/collapser.py (hunk 3 above)
Comment (andrewtarzia) [reply stko/optimizers/collapser.py:165]: Not every function requires docs is the main reason

Comment (andrewtarzia) [reply stko/optimizers/collapser.py:165]: That said, some of these doc strings could improve and have been improved

File: stko/optimizers/collapser.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+        return min_dist
+
+    def _no_short_contacts(self, mol):
+        """
+        Define a function that determines when collapsing should stop.
+
+        """
+
+        for dist in self._get_inter_BB_distance(mol):
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply stko/optimizers/collapser.py:142]: does not quite do as intended.

[the remark this answers is not in the document]
Comment (andrewtarzia) [reply stko/optimizers/collapser.py:142]: `any(dist < self._distance_cut for dist in self._get_inter_BB_distance(mol))`

[the remark this answers is not in the document]
Comment (andrewtarzia) [reply stko/optimizers/collapser.py:142]: I revert my statement

File: stko/optimizers/collapser.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+                yield dist
+
+    def _get_min_inter_BB_distance(self, mol):
+        """
+        Calculate the minimum inter building block distance.
+
+        """
+
+        min_dist = 1000
+        for dist in self._get_inter_BB_distance(mol):
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply stko/optimizers/collapser.py:131]: no idea.

**Pull request merged**